### PR TITLE
Replace a recursive algorithm with an iterative one and a stack.

### LINKED ISSUE #74931: "thread 'rustc' has overflowed its stack" introduced in 1.45.0
### Code

```bash
git clone https://github.com/crate-ci/typos.git
cd typos/crates/typos-dict
cargo +1.44.0 check  # works
cargo +1.45.0 check  # overflows
RUST_MIN_STACK=20000000 cargo check  # works
```
(also reproduced in beta and nightly)

### Meta

`rustc --version --verbose`:
```
rustc 1.45.0 (5c1f21c3b 2020-07-13)
binary: rustc
commit-hash: 5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2
commit-date: 2020-07-13
host: x86_64-apple-darwin
release: 1.45.0
LLVM version: 10.0
```

### Error output

```
Checking typos-dict v0.2.1 (/Users/edpage/src/personal/typos/crates/typos-dict)

thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
error: could not compile `typos-dict`.

Caused by:
  process didn't exit successfully: `rustc --crate-name typos_dict --edition=2018 crates/typos-dict/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata -Cembed-bitcode=no -C debuginfo=2 -C metadata=439c5583e8d5e9ec -C extra-filename=-439c5583e8d5e9ec --out-dir /Users/edpage/src/personal/typos/target/debug/deps -C incremental=/Users/edpage/src/personal/typos/target/debug/incremental -L dependency=/Users/edpage/src/personal/typos/target/debug/deps --extern log=/Users/edpage/src/personal/typos/target/debug/deps/liblog-6af237d64da748f9.rmeta --extern phf=/Users/edpage/src/personal/typos/target/debug/deps/libphf-3c962e541346fbff.rmeta --extern unicase=/Users/edpage/src/personal/typos/target/debug/deps/libunicase-561e4e81056a68fd.rmeta` (signal: 6, SIGABRT: process abort signal)
```

<details><summary><strong>Backtrace</strong></summary>
<p>

No backtrace generated

</p>
</details>

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #74931

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

Seems not unlikely to have some perf effects.

File: src/librustc_mir/transform/simplify.rs [hunk 1]
```diff
@@ -172,9 +173,8 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
         }
     }
 
-    // Collapse a goto chain starting from `start`
-    fn collapse_goto_chain(&mut self, start: &mut BasicBlock, changed: &mut bool) {
-        let mut terminator = match self.basic_blocks[*start] {
+    fn extract_terminator(&mut self, bb: BasicBlock) -> Option<Terminator<'tcx>> {
```
Comment (ecstatic-morse) [inline src/librustc_mir/transform/simplify.rs:176]: Can we use a more descriptive name for this? `take_terminator_if_simple_goto` or so?

File: src/librustc_mir/transform/simplify.rs [hunk 2]
```diff
@@ -183,32 +183,43 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
...
-        *start = target;
+            if self.pred_count[current] == 1 {
+                // This is the last reference to current, so the pred-count to
+                // to target is moved into the current block.
+                self.pred_count[current] = 0;
+            } else {
+                self.pred_count[*target] += 1;
+                self.pred_count[current] -= 1;
+            }
+            *changed |= current != *target;
```
Comment (ecstatic-morse) [inline src/librustc_mir/transform/simplify.rs:220]: Preexisting I know, but why do we only mark the MIR as changed if `current != target`? If `current == target` then we are inserting a self-loop, but it's possible for this to represent a change in the MIR if we had a cycle involving several `Goto`s.

File: src/librustc_mir/transform/simplify.rs [hunk 3]
```diff
@@ -183,32 +183,43 @@ impl<'a, 'tcx> CfgSimplifier<'a, 'tcx> {
...
-        if self.pred_count[*start] == 1 {
-            // This is the last reference to *start, so the pred-count to
-            // to target is moved into the current block.
-            self.pred_count[*start] = 0;
-        } else {
-            self.pred_count[target] += 1;
-            self.pred_count[*start] -= 1;
+    // Collapse a goto chain starting from `start`
+    fn collapse_goto_chain(&mut self, start: &mut BasicBlock, changed: &mut bool) {
+        let mut terminators: SmallVec<[_; 1]> = Default::default();
```
Comment (ecstatic-morse) [inline src/librustc_mir/transform/simplify.rs:192]: Why the `SmallVec` here? In the common case, `start` will not refer to a `Goto` with no statements, so `terminators` will be empty.

**Review by ecstatic-morse — COMMENTED** [review]
Thanks for doing this @oli-obk. I remember looking at this code early on and wondering if it was going to ruin someone's day :smile:. This is a pretty straightforward translation of the recursive version, so I feel good about it even though I left some questions.

File: src/librustc_mir/transform/simplify.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/transform/simplify.rs:220]: I have learned the hard way that tinkering with this code is not a good idea XD But I've also learned that you either break it completely or not at all, there's no middle ground, so I guess i can do more experimentation

File: src/librustc_mir/transform/simplify.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_mir/transform/simplify.rs:192]: anecdotal evidence showed me many single-step goto chains (I had logging on during libcore builds). But we can also just go ahead and benchmark `Vec` vs `SmallVec<[_; 1]>` and maybe some other default sizes.

File: src/librustc_mir/transform/simplify.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/transform/simplify.rs]: oh well, it worked :D

Comment (ecstatic-morse) [reply src/librustc_mir/transform/simplify.rs]: If you don't wanna lump it into this PR, that's fine too. It just doesn't quite make sense to me.

File: src/librustc_mir/transform/simplify.rs (hunk 3 above)
Comment (ecstatic-morse) [reply src/librustc_mir/transform/simplify.rs:196]: Sounds good. It shouldn't matter much either way, maybe leave a comment?

**Comment by ecstatic-morse** [comment]
@bors rollup

@oli-obk r=me when you're happy

File: src/librustc_mir/transform/simplify.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/transform/simplify.rs]: it's already in this PR ;)

**Pull request merged**